src/NServiceBus.Core/Recoverability/CustomAction.cs

```diff
@@ -0,0 +1,33 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Transport;
+
+    /// <summary>
+    /// Defines a custom recoverability action, taking care of handling a message processing error.
+    /// </summary>
+    public class CustomAction : RecoverabilityAction
+    {
+        readonly Func<ErrorContext, Task> customAction;
+
+        /// <summary>
+        /// Creates a new <see cref="CustomAction"/> using the provided callback to resolve a message failure.
+        /// </summary>
+        /// <param name="customAction">The method invoked to handle a failed message.</param>
+        public CustomAction(Func<ErrorContext, Task> customAction)
+        {
+            this.customAction = customAction;
+        }
+
+        /// <summary>
+        /// Invokes the custom recoverability action.
+        /// </summary>
+        /// <returns>always returns <see cref="ErrorHandleResult.Handled"/> to indicate the transport to consume the message.</returns>
+        public async Task<ErrorHandleResult> Invoke(ErrorContext errorContext)
```

Comment:
I assume this is public to enable testing?

Reply:
yeah that was the intention, didn't spend too much time thinking about it though

Reply:
Sent in an addition https://github.com/Particular/NServiceBus/pull/5608